scapy/contrib/rpl.py

```diff
@@ -0,0 +1,312 @@
+                   ByteField("reserved", 0),
+                   IP6Field("dodagid", "::1")]
+
+
+class _OptDODAGIDField(IP6Field):
+    """
+    Handle Optional DODAG ID field in DAO
+    """
+    def getfield(self, pkt, s):
+        if pkt.D == 0:
+            return s, None
+        return s[16:], self.m2i(pkt, s[:16])
+
+    def addfield(self, pkt, s, val):
+        if pkt.D == 1:
+            return s + self.i2m(pkt, val)
+        if val:
+            print("RPL DAO 'D' flag is not set but dodagid is given.")
+        return s
+
+
+class DAO(_RPLGuessOption, Packet):
+    """
+    Control Message: Destination Advertisement Object (DAO)
+    """
+    name = "Destination Advertisement Object"
+    fields_desc = [ByteField("RPLInstanceID", 50),
+                   BitField("K", 0, 1),
+                   BitField("D", 0, 1),
+                   BitField("flags", 0, 6),
+                   ByteField("reserved", 0),
+                   ByteField("daoseq", 1),
+                   _OptDODAGIDField("dodagid", None)]
+
+
+class DAOACK(_RPLGuessOption, Packet):
+    """
+    Control Message: Destination Advertisement Object Acknowledgement (DAOACK)
+    """
+    name = "Destination Advertisement Object Acknowledgement"
+    fields_desc = [ByteField("RPLInstanceID", 50),
+                   BitField("D", 0, 1),
+                   BitField("reserved", 0, 7),
+                   ByteField("daoseq", 1),
+                   ByteField("status", 0),
+                   _OptDODAGIDField("dodagid", None)]
+
+
+# https://datatracker.ietf.org/doc/draft-ietf-roll-efficient-npdao/
+class DCO(_RPLGuessOption, Packet):
+    """
+    Control Message: Destination Cleanup Object (DCO)
+    """
+    name = "Destination Cleanup Object"
+    fields_desc = [ByteField("RPLInstanceID", 50),
+                   BitField("K", 0, 1),
+                   BitField("D", 0, 1),
+                   BitField("flags", 0, 6),
+                   ByteField("status", 0),
+                   ByteField("dcoseq", 1),
+                   _OptDODAGIDField("dodagid", None)]
+
+
+# https://datatracker.ietf.org/doc/draft-ietf-roll-efficient-npdao/
+class DCOACK(_RPLGuessOption, Packet):
+    """
+    Control Message: Destination Cleanup Object Acknowledgement (DCOACK)
+    """
+    name = "Destination Cleanup Object Acknowledgement"
+    fields_desc = [ByteField("RPLInstanceID", 50),
+                   BitField("D", 0, 1),
+                   BitField("flags", 0, 7),
+                   ByteField("dcoseq", 1),
+                   ByteField("status", 0),
+                   _OptDODAGIDField("dodagid", None)]
+
+
+# https://www.iana.org/assignments/rpl/rpl.xhtml#control-codes
+bind_layers(RPL, DIS, code=0)
+bind_layers(RPL, DIO, code=1)
+bind_layers(RPL, DAO, code=2)
+bind_layers(RPL, DAOACK, code=3)
+bind_layers(RPL, DCO, code=7)
+bind_layers(RPL, DCOACK, code=8)
+
+
+# https://www.iana.org/assignments/rpl/rpl.xhtml#control-message-options
+for msg in [DIS, DIO, DAO, DAOACK, DCO, DCOACK]:
```

Comment:
You can use `bind_layers(_RPLGuessOption, ...)` instead of doing it for each since they all inherit from it

Reply:
Actually there are two things:
1. RPL Control Messages (DIS/DIO/DAO/DAO...)
2. RPL Control Options (Pad1/N, Target, TIO, PIO, RIO ...)

The messages can happen one or more options. So I need DIS/DIO.. to be represented by a different dummy class. I am now using _RPLGuessMsgType to represent Messages ... and the previously used _RPLGuessOption for options.
Kindly check if this makes sense.